stream: add destroy and _destroy methods.

mcollina opened the pull request:
Adds destroy() and _destroy() methods to Readable, Writable, Duplex
and Transform. It also standardize the behavior and the implementation
of destroy(), which has been incosistent in userland.
This PR also updates all subsystems of core to use the new destroy().

Fixes: https://github.com/nodejs/readable-stream/issues/124

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)

stream, net, fs, process

mcollina commented:
cc @nodejs/streams

mcollina commented:
This change might be semver-major, because now we are always emitting the error passed to `destroy` asynchronously, but stdio did so synchronously: https://github.com/nodejs/node/pull/12925/files#diff-5224cdaa6225b9d4e1e7bd73a2ea284f.
However, that might be also seen as a bug of the previous implementation, as all the other parts of core emitted those asynchronously.

cjihrig commented:
This has a conflict in doc/api/stream.md. We'll definitely want a CITGM run on this too.

thefourtheye commented:
This already has a merge conflict 😮

thefourtheye commented on doc/api/stream.md:
```diff
@@ -499,6 +499,13 @@ write('hello', () => {
 
+##### writable.destroy([error])
+
+destroy the stream, and emit the passed error. after this call, the
```
Initcap please.

cjihrig commented on doc/api/stream.md:
```diff
@@ -499,6 +499,13 @@ write('hello', () => {
 
 A Writable stream in object mode will always ignore the `encoding` argument.
 
+##### writable.destroy([error])
```
The newly added APIs should probably have changelogs.

cjihrig commented on doc/api/stream.md:
```diff
@@ -1359,6 +1382,12 @@ The `writable._writev()` method is prefixed with an underscore because it is
+#### writable.\_destroy(err, callback)
+
+* `err` {Error} An error.
+* `callback` {Function} A callback function (optionally with an error
```
A callback function that takes an optional error argument which is invoked when the writable is destroyed.

cjihrig commented on lib/_stream_readable.js:
```diff
@@ -132,9 +135,25 @@ function Readable(options) {
   if (options && typeof options.read === 'function')
     this._read = options.read;
 
+  if (options && typeof options.destroy === 'function')
```
Can `options` be `typeof options === 'object'`. I believe by this point we know that it isn't `null`.

calvinmetcalf replied:
we do the same thing 3 lines earlier so so consistency wise it's probably fine.

cjihrig commented on test/parallel/test-process-external-stdio-close-spawn.js:
```diff
@@ -0,0 +1,26 @@
+    process.disconnect();
+  }));
+} else {
+  // '--inspect', '--inspect-brk',
```
This can be removed.

cjihrig commented on test/parallel/test-process-external-stdio-close-spawn.js:
```diff
@@ -0,0 +1,26 @@
+  }));
+} else {
+  // '--inspect', '--inspect-brk',
+  const child = cp.spawn(process.execPath, [ __filename, 'child'], {
```
Unnecessary space before `__filename`

cjihrig commented on test/parallel/test-stream-duplex-destroy.js:
```diff
@@ -0,0 +1,158 @@
+    read() {}
+  });
+
+  duplex._destroy = function(err, cb) {
```
Might be worth wrapping the right hand side in `common.mustCall()` throughout if you know exactly how many times it will be called.

jasnell commented on doc/api/stream.md:
```diff
@@ -1073,6 +1080,13 @@ myReader.on('readable', () => {
+
+Destroy the stream, and emit the passed error. After this call, the
+readable stream would release any internal resources.
+implementors should not override this method, but implement to
```
`s/implement to/implement`

calvinmetcalf commented on lib/_stream_transform.js:
```diff
@@ -194,6 +194,14 @@ Transform.prototype._read = function(n) {
 
 
+Transform.prototype._destroy = function(err, cb) {
+  Duplex.prototype._destroy.call(this, err, (err2) => {
```
can we just use `super._destroy` here ?

mcollina replied:
we are not inheriting with `class`, so no `super`.

calvinmetcalf replied:
I don't think that matters

calvinmetcalf replied:
right only for object shorthand syntax my bad.

mcollina commented:
@cjihrig please have a look.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/7989/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/768/
CITGM with `READLE_STREAM=disable`:  https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/770/

CITGM from master: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/771/

mcollina commented:
@MylesBorins can you look at those citgm failures from master? There seems to be quite a bit of breakage even on the one on master, and I have trouble reading out if I broke something here.
Also, there were a bit of npm failures as well.

gibfahn commented:
For CitGM, `winston` and `watchify` failures are expected, see https://github.com/nodejs/node/pull/12220#issuecomment-300366441

mcollina commented:
So, we are breaking a bit the ecosystem for this, here are duplexify tests:

```
$ READABLE_STREAM=disable ../node/node test.js
TAP version 13
# passthrough
ok 1 should finish
ok 2 same in as out
# passthrough + double end
ok 3 should finish
ok 4 same in as out
# async passthrough + end
ok 5 should finish
ok 6 same in as out
# duplex
ok 7 onwrite should match
ok 8 onwrite should match
ok 9 onwrite should match
ok 10 should finish
ok 11 ondata should match
ok 12 ondata should match
ok 13 ondata should match
ok 14 should end
# async
ok 15 same in as out
# destroy
ok 16 write destroyed
ok 17 close emitted
ok 18 write destroyed
not ok 19 plan != count
  ---
    operator: fail
    expected: 2
    actual:   3
    at: Transform.write.destroy (/Users/matteo/Repositories/duplexify/test.js:126:7)
  ...
ok 20 close emitted
# destroy both
ok 21 read destroyed
ok 22 write destroyed
ok 23 close emitted
ok 24 read destroyed
not ok 25 plan != count
  ---
    operator: fail
    expected: 3
    actual:   4
    at: Transform.read.destroy (/Users/matteo/Repositories/duplexify/test.js:149:7)
  ...
ok 26 write destroyed
ok 27 close emitted
# bubble read errors
ok 28 received read error
ok 29 close emitted
not ok 30 received read error
  ---
    operator: deepEqual
    expected: 'read-error'
    actual:   'write-error'
    at: Duplexify.<anonymous> (/Users/matteo/Repositories/duplexify/test.js:168:7)
  ...
not ok 31 plan != count
  ---
    operator: fail
    expected: 2
    actual:   3
    at: Duplexify.<anonymous> (/Users/matteo/Repositories/duplexify/test.js:168:7)
  ...
ok 32 close emitted
# bubble write errors
ok 33 received write error
ok 34 close emitted
not ok 35 received write error
  ---
    operator: deepEqual
    expected: 'write-error'
    actual:   'read-error'
    at: Duplexify.<anonymous> (/Users/matteo/Repositories/duplexify/test.js:186:7)
  ...
not ok 36 plan != count
  ---
    operator: fail
    expected: 2
    actual:   3
    at: Duplexify.<anonymous> (/Users/matteo/Repositories/duplexify/test.js:186:7)
  ...
ok 37 close emitted
# reset writable / readable
ok 38 should be equivalent
ok 39 should be equivalent
ok 40 should be equivalent
# cork
ok 41 (unnamed assert)
# prefinish not twice
ok 42 only prefinish once
# close
ok 43 should forward close
ok 44 should forward close
not ok 45 .end() called twice
  ---
    operator: fail
    at: Duplexify.<anonymous> (/Users/matteo/Repositories/duplexify/test.js:268:7)
  ...
# works with node native streams (net)
ok 46 should be equivalent

1..46
# tests 46
# pass  39
# fail  7
```

calvinmetcalf commented:
what exactly is the issue with destory? subtly different mechanics ?

mcollina commented:
> what exactly is the issue with destory? subtly different mechanics ?

@calvinmetcalf me breaking things. `destroyed`  has no setter, and it's used in the ecosystem.
I should have fixed that, let's see.

CITGM (disabling readable-stream): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/774/

mcollina commented:
new run: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/785/

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8083/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/786/

(fixed a regression in `from2`).

jasnell commented on doc/api/stream.md:
```diff
@@ -499,6 +499,16 @@ write('hello', () => {
 
+##### writable.destroy([error])
+<!-- YAML
+added: TBD
```
s/TBD/REPLACEME

jasnell commented on doc/api/stream.md:
```diff
@@ -1073,6 +1083,16 @@ myReader.on('readable', () => {
 
+##### readable.destroy([error])
+<!-- YAML
+added: TBD
```
s/TBD/REPLACEME

jasnell replied:
(I'll stop pointing these out.. :-) ...)

jasnell commented on lib/_stream_duplex.js:
```diff
@@ -76,3 +76,37 @@ function onend() {
 }
+
+Object.defineProperty(Duplex.prototype, 'destroyed', {
+  get: function() {
```
nit: simply to just `get() { ... }`

jasnell commented on lib/_stream_duplex.js:
```diff
@@ -76,3 +76,37 @@ function onend() {
+
+  if (err) {
+    this.emit('error', err);
+  }
```
should this return after emitting the error?

mcollina replied:
In fact this piece is just wrong. We need to call the callback with the error here. The net effect is that `'err'` will be emitted in the next tick.

jasnell commented on lib/_stream_readable.js:
```diff
@@ -129,12 +132,44 @@ function Readable(options) {
+});
+
+Readable.prototype.destroy = require('internal/streams/destroy').destroy;
+Readable.prototype._undestroy = require('internal/streams/destroy').undestroy;
```
require once..?

jasnell commented on lib/_stream_writable.js:
```diff
@@ -563,3 +569,30 @@ function onCorkedFinish(corkReq, state, err) {
+});
+
+Writable.prototype.destroy = require('internal/streams/destroy').destroy;
+Writable.prototype._undestroy = require('internal/streams/destroy').undestroy;
```
require once?

jasnell commented on test/parallel/test-process-external-stdio-close-spawn.js:
```diff
@@ -0,0 +1,28 @@
+if (process.argv[2] === 'child') {
+  process.on('message', common.mustCall((msg) => {
+    assert.strictEqual(msg, 'go');
+    console.log('logging should not cause a crash');
```
nit: would prefer to eliminate the extraneous `console.log` output

mscdex replied:
I agree.

mcollina replied:
the `console.log` message is what will cause the test to fail in case of a regression. It is the central part of the test. This is the approach in the test from which this was inspired [test/parallel/test-process-external-stdio-close.js](https://github.com/mcollina/node/blob/a04c7313f298f123fc48fc48cd57dc460addc2a7/test/parallel/test-process-external-stdio-close.js) by @addaleax.

cjihrig replied:
Can you add a comment stating that the `console.log()` is part of the test.

mscdex commented:
There are some typos in the commit message FWIW.

mcollina commented:
Can someone from @nodejs/citgm confirm those failures are all expected?

mcollina commented:
Rebased and fixed the nits.

CI: https://ci.nodejs.org/job/node-test-pull-request/8107/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/790/

cc @jasnell @cjihrig

mscdex commented:
s/incosistent/inconsistent/ in the commit message.

toboid commented on test/parallel/test-stream-duplex-destroy.js:
```diff
@@ -0,0 +1,195 @@
+  duplex.on('end', common.mustCall());
+  duplex.on('finish', common.mustCall());
+  assert.strictEqual(duplex.destroyed, true);
+  assert.strictEqual(duplex.destroyed, true);
```
These 2 assertions are identical

toboid commented on test/parallel/test-stream-duplex-destroy.js:
```diff
@@ -0,0 +1,195 @@
+  duplex.on('end', common.mustNotCall('no end event'));
+  duplex.on('finish', common.mustNotCall('no finish event'));
+
+  // error is swalled by the custom _destroy
```
Little typo here - *swalled* instead of *swallowed*

mcollina commented:
@toboid @mscdex @jasnell @cjihrig things should be ok now.

cjihrig commented on test/parallel/test-stream-writable-destroy.js:
```diff
@@ -0,0 +1,172 @@
+  write._destroy = common.mustCall(function(err, cb) {
+    assert.strictEqual(err, null);
+    cb(expected);
+  }, 1);
```
The 1 isn't really necessary here and in other places. `common.mustCall()` defaults to 1.

cjihrig commented on doc/api/stream.md:
```diff
@@ -499,6 +499,16 @@ write('hello', () => {
+
+Destroy the stream, and emit the passed error. After this call, the
+writible stream has ended.
+Implementors should not override this method, but instead implement
```
The previous line seems short. Could some of this line be moved up?

mcollina commented:
@cjihrig updated.

mcollina commented:
CITGM failures are currently blocked by https://github.com/nodejs/node/issues/13045.

refack commented:
> CITGM failures are currently blocked by #13045 cc @mscdex, @refack.

There's a fix in the pipeline #13092

mcollina commented:
@refack when that lands, I'll rebase and test with the ecosystem.

jasnell commented:
looking at it again, this is definitely semver-major due to changes in the error handling when attempting to destroy stdout/stderr.

jasnell commented:
@mcollina ... I'm going to be refreshing the 8.x branch on Monday or Tuesday and will be doing another RC build on Tuesday. We're just over a week from the release. Even as a semver-major, I want to get this into 8.0.0 so let's definitely get it landed before I do the next RC.

refack commented:
@mcollina #13092 landed.

mcollina commented:
Rebased.
CI: https://ci.nodejs.org/job/node-test-pull-request/8205/
CITGM (disabled readable-stream): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/798/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/800/
CITGM (master): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/801/

mcollina commented:
This can be landed as CITGM is good (same failures as on master). I'll land on Monday morning CEST.

mcollina commented:
Landed as 330c8d7

refack commented:
Sorry I mixed up this and #12828 
Intended original comment:

---
Probably broke `ws` https://github.com/websockets/ws/issues/1118
/cc @lpinca

MylesBorins commented:
this technically landed after the semver major cutoff @nodejs/ctc does anyone have an issue with this landing in the 8.0.0 release?

The fact that it may be breaking ws has me slightly concerned

mcollina commented:
https://github.com/nodejs/node/pull/13156 needs to land as well to fix the ws regression.

jasnell commented:
I'm definitely +1 on this landing in 8.0.0
